content/matrices/components/gauss-solver.ts

```diff
@@ -0,0 +1,196 @@
+              }
+              break;
+            case 'add':
+              // "Add a multiple of one row to another"
+              val.text = this.$inputCells[i][j].text;
+              if (state.factor && i === state.outRow1) {
+                const c1 = Expression.parse(this.$inputCells[state.inRow1][j].text).evaluate();
+                const c2 = Expression.parse(this.$inputCells[state.inRow2][j].text).evaluate();
+                val.text = '' + (c1 + state.factor * c2);
+              } else {
+                val.text = this.$inputCells[i][j].text;
+              }
+              break;
+            case 'swap':
+              // "Swap two rows"
+              if (i === state.outRow1) {
+                val.text = this.$inputCells[state.inRow1][j].text;
+              } else if (i === state.outRow2) {
+                val.text = this.$inputCells[state.inRow2][j].text;
+              } else {
+                val.text = this.$inputCells[i][j].text;
+              }
+              break;
+          }
+        }
+      }
+    });
+
+    // Button to Apply Row Operation
+    const $applyBtn = this.$('.apply');
+    $applyBtn?.on('click', () => {
+      // swapping input and output
+      for (const [i, row] of this.$inputCells.entries()) {
+        for (const [j, val] of row.entries()) {
+          val.text = this.$outputCells[i][j].text;
+        }
+      }
+      this.inputStack.push(this.$outputCells.map(row => row.map(val => val.text)));
+      console.log(this.inputStack);
+
+      if (this.checkForSolvedIdentity()) {
+        console.log('Success!');
+        this.$step.tools.confetti();
+      }
+    });
+
+    const $undoBtn = this.$('.undo');
+    $undoBtn?.on('click', () => {
+      if (this.inputStack.length === 1) return;
+      this.inputStack.pop();
+
+      const newTop = this.inputStack[this.inputStack.length - 1];
+      for (const [i, row] of this.$inputCells.entries()) {
+        for (const [j, val] of row.entries()) {
+          val.text = newTop[i][j];
+        }
+      }
+    });
+
+    // watch for factor change
+    this.model.watch((state) => {
+      let _parseable = true;
+      let expr;
+      let value;
+      try {
+        // FIXME: @philipp this might throw an error, Not sure how else to handle it.
+        expr = Expression.parse(state.factorString);
+        value = expr.evaluate() as number;
+        this.model.factor = value;
+      } catch (e) {
+        _parseable = false;
+        this.model.factor = 1;
+      }
+    });
+  }
+
+  checkForSolvedIdentity(): boolean {
+    console.log('Checking for solved identity');
+    const numRows = this.$inputCells.length;
+
+    for (let i = 0; i < numRows; i++) {
+      const values = this.$inputCells[i];
+      // i is also the expectedOneIndex
+
+      // check all values
+      for (let j = 0; j < numRows; j++) {
+        // console.log(`Checking row=${i}; col=${j}; value=${values[j]}`);
+        if (j === i) {
+          if (values[j].text != '1') {
```

Comment:
We should avoid storing state in the DOM. Rather than accessing `.text` (which is quite an expensive operation), we should just have a numeric, nested array that contains the current value of the output matrix